test/hotspot/jtreg/compiler/locks/TestNestedLocksElimination.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8268347
+ * @summary Nested locks optimization may create unbalanced monitor enter/exit code
+ *
+ * @run main/othervm -XX:-BackgroundCompilation
+ *                   -XX:CompileCommand=dontinline,TestNestedLocksElimination::foo
+ *                   -XX:CompileCommand=dontinline,TestNestedLocksElimination::getNext
+ *                   -XX:CompileCommand=dontinline,TestNestedLocksElimination::getHolder
+ *                   TestNestedLocksElimination
+ */
+
+import java.util.LinkedList;
+
+public class TestNestedLocksElimination {
+
+    private LinkedList<char[]> buffers = new LinkedList<>();
+    private boolean complete = false;
+    private int bufferSize;
+
+    void foo(char[] ca) {
+        // Don't inline dummy method
+    }
+
+    char[] getNext(int length, int count) {
+        // Don't inline
+        if (this.buffers.isEmpty()) {
+            return new char[100];
+        }
+        char[] b = (char[]) this.buffers.getFirst();
+        if (count >= 100) {
+            this.complete = true;
+            this.buffers.clear(); // empty
+        }
+        return b;
+    }
+
+    synchronized boolean isComplete() {
+        return this.complete;
+    }
+
+    synchronized boolean availableSegment() {
+        return (buffers.isEmpty() == false);
+    }
+
+    TestNestedLocksElimination getHolder(TestNestedLocksElimination s1, TestNestedLocksElimination s2, int count) {
```

Comment:
Should have this comment here also:

    `// Don't inline`

Reply:
Will do.